feat: Epitaph Road (Rathnite Foothills)

pacampbell opened the pull request:
Added support for Epitaph Road endgame content.
- Implemented the personal quest to enter "Heroic Spirit Sleeping Path:
  Rathnite Foothills".
- Created new epitaph assets which describe the different paths,
  sections and trials.
- Implemented multiple packet handlers for different season dungeon
   packets.
- Implemented queued packet handler which enables packets to be send
   after the response is sent.
- Implemented sections 1, 2, 3 and 4 for Rathnite Foothills
- Implemented the final trial for Rathnite Foothills
- Implemented sections 1 and 3 for Memory of Megadosys.

![image](https://github.com/user-attachments/assets/50f1cbf8-6157-4dbf-aefd-dda15fbe38a7)

# Checklist:
- [x] The project compiles
- [x] The PR targets `develop` branch

pacampbell committed:
implement new packets, handlers and database queries

- Implemented many req, res, ntc and handlers for Epitaph content.
- Renamed existing BonusDungeonManager to DungeonManager.
- Implemented database migration from 24 -> 25.
- Implemented a new packet handler which provides a queue which can be
  used to send packets after the response is returned.
- Commented out RPC ping print message.
- Added new NPC behavior to NpcGetExtendedFacilityHandler.

pacampbell committed:
Add epitaph required assets

- Created personal quest required to enter Rathnite Foothills dungeon.
- Created new EpitaphRoad.json file which contains dungeon information
  for each path.
- Created new epitaph directory to hold all trials
  - Implemented all trials for Rathnite foothills.
  - Implemented section 1 and section 3 trials for Memory of Megadosys.

pacampbell committed:
docs: Add new docs for epitaph

- Added docs which contain information found about buffs, objectives and rewards.
- Started a new document guide.md which describes the epitaph assets.

pacampbell committed:
enemyspawn: Add enemies for Rathnite foothills

- Added new enemies for the Rathnite Foothills dungeon.
- Added some enemies to memory of megadosys.

pacampbell commented on Arrowgene.Ddon.Rpc/RpcServer.cs:
```diff
@@ -21,7 +21,7 @@ public RpcServer(DdonGameServer gameServer)
         {
             RpcCommandResult result = command.Execute(_gameServer);
-            Logger.Info(result.ToString());
+            // Logger.Info(result.ToString());
```
This log spams so often, it gets in the way of reading the logs when ddon-tools is connected. We should make some filter system for this so we only get logs we want.

alborrajo replied:
Not against removing this comment. It's true it gets spammy and i don't think we're getting much info out of it

pacampbell commented on Arrowgene.Ddon.GameServer/Characters/CharacterManager.cs:
```diff
@@ -59,6 +59,14 @@ public Character SelectCharacter(uint characterId, DbConnection? connectionIn =
                 return null;
             }
 
+            character.EpitaphRoadState.UnlockedContent = _Server.Database.GetEpitaphRoadUnlocks(character.CharacterId);
```
I'm not sure what you mean by this. Every character needs the details populated. Old logic should be moved into the character manager instead of the way it was originally designed so we can move the transaction logic into these functions.

pacampbell replied:
This function is called when you log in.

pacampbell replied:
I see it got used in some of these pawn ones. Maybe we should make a lighter ones for those? Problem is pawns need the character inventory and all that other unpleasant stuff.

pacampbell committed:
party fixes

- Add cancel handler
- Fix bug when party ready/unready logic
- Add missing connectionIn to DB queries
- Fix argument id in om command
- Fix issue where repop NTC was only sent to the player who killed the
  enemy.
- Changed certain OM update packets to only send to the client who
  requested.
- Fix barrier purchase so it only unlocks for the person who bought.

alborrajo commented on Arrowgene.Ddon.Shared/Entity/PacketStructure/C2SSeasonDungeonSoulOrdealReadyReq.cs:
```diff
@@ -0,0 +1,39 @@
+    public class C2SSeasonDungeonSoulOrdealReadyReq : IPacketStructure
+    {
+        public C2SSeasonDungeonSoulOrdealReadyReq()
+        {
```
Should instance StageLayoutId in the constructor

alborrajo commented on Arrowgene.Ddon.Shared/Entity/PacketStructure/C2SSeasonDungeonUpdateKeyPointDoorStatusReq.cs:
```diff
@@ -0,0 +1,39 @@
+
+        public PacketId Id => PacketId.C2S_SEASON_DUNGEON_UPDATE_KEY_POINT_DOOR_STATUS_REQ;
+
+        public CDataStageLayoutId StageLayoutId { get; set; }
```
Should instance StageLayoutID in the constructor

alborrajo commented on Arrowgene.Ddon.Shared/Entity/PacketStructure/C2S_SEASON_62_40_16_NTC.cs:
```diff
@@ -0,0 +1,36 @@
+        /// </summary>
+        public PacketId Id => PacketId.C2S_SEASON_62_40_16_NTC;
+
+        public CDataStageLayoutId LayoutId { get; set; }
```
Should instance LayoutId in the constructor

alborrajo commented on Arrowgene.Ddon.Shared/Entity/PacketStructure/C2S_SEASON_DUNGEON_62_12_16_NTC.cs:
```diff
@@ -0,0 +1,31 @@
+    {
+        public PacketId Id => PacketId.C2S_SEASON_DUNGEON_62_12_16_NTC;
+
+        public CDataStageLayoutId LayoutId { get; set; }
```
Should instance LayoutId in the constructor

alborrajo commented on Arrowgene.Ddon.Shared/Entity/PacketStructure/S2CSeasonDungeonReceiveSoulOrdealRewardRes.cs:
```diff
@@ -0,0 +1,35 @@
+        public override PacketId Id => PacketId.S2C_SEASON_DUNGEON_RECEIVE_SOUL_ORDEAL_REWARD_RES;
+
+        public S2CSeasonDungeonReceiveSoulOrdealRewardRes()
+        {
```
Should instance RewardList in the constructor

alborrajo commented on Arrowgene.Ddon.Shared/Entity/PacketStructure/S2CSeasonDungeonSetOmStateNtc.cs:
```diff
@@ -0,0 +1,35 @@
+    {
+        public PacketId Id => PacketId.S2C_SEASON_DUNGEON_SET_OM_STATE_NTC;
+
+        public CDataStageLayoutId LayoutId { get; set; }
```
Should instance LayoutId in the constructor

alborrajo commented on Arrowgene.Ddon.Shared/Entity/PacketStructure/S2CStageGetSpAreaChangeInfoRes.cs:
```diff
@@ -0,0 +1,52 @@
+
+
+        public S2CStageGetSpAreaChangeInfoRes()
+        {
```
Should instance Unk1 and EntryCostList in the constructor

alborrajo commented on Arrowgene.Ddon.Shared/Entity/Structure/CDataSeasonDungeonBuffEffectReward.cs:
```diff
@@ -0,0 +1,39 @@
+    public class CDataSeasonDungeonBuffEffectReward
+    {
+        public CDataSeasonDungeonBuffEffectReward()
+        {
```
Should instance BuffName in the constructor

alborrajo reviewed (commented):
These features are getting crazy complicated, good job on it.
I haven't focused too much on the packets, managers, and handlers as i trust that they work, the comments im giving are mostly nitpicks, places where i feel are missing comments or explaining where certain values come from, and suggestions that you can freely ignore if you wish.

pacampbell commented on Arrowgene.Ddon.GameServer/Handler/NpcGetExtendedFacilityHandler.cs:
```diff
@@ -16,12 +21,85 @@ public NpcGetExtendedFacilityHandler(DdonGameServer server) : base(server)
+                    result.ExtendedMenuItemList.Add(new CDataNpcExtendedFacilityMenuItem() { FunctionClass = NpcFunction.WarMissions, FunctionSelect = NpcFunction.GiveSpirits });
+                }
+            }
+#if false
```
I just disabled it for now since I didn't finish everything in that area yet.

alborrajo replied:
Please leave a TODO comment explaining that :pray:

pacampbell commented on Arrowgene.Ddon.GameServer/Handler/StageGetSpAreaChangeInfoHandler.cs:
```diff
@@ -0,0 +1,25 @@
+            {
+                Unk0 = 19,
+                Unk1 = "Memory of Megadosys",
+                StageId = 559,
```
I forgot about these ones. I made some changes and now these handlers don't even get called. I should probably delete this stuff so it makes some error or doesn't nothing instead of warping the player.

pacampbell commented on Arrowgene.Ddon.GameServer/Characters/EpitaphRoadManager.cs:
```diff
@@ -0,0 +1,1324 @@
+                    // result as seen in player videos
+                    results.Add(new InstancedGatheringItem()
+                    {
+                        ItemId = (gatheringPoint == null) ? 9393 : dungeonInfo.SoulItemId,
```
I suppose this could be more robust than it currently is.

pacampbell commented on Arrowgene.Ddon.GameServer/GatheringItems/InstanceEpitaphRoadDropItemManager.cs:
```diff
@@ -0,0 +1,122 @@
+            return 0;
+        }
+
+        private readonly List<(EpitaphBuffId BuffId, uint ItemId, uint Amount, double Chance)> gDropConfigs = new List<(EpitaphBuffId BuffId, uint ItemId, uint Amount, double Chance)>()
```
They could be configurable, will add a TODO

alborrajo commented on Arrowgene.Ddon.GameServer/Chat/Command/Commands/UpdateOmCommand.cs:
```diff
@@ -0,0 +1,83 @@
+{
+    public class UpdateOmCommand : ChatCommand
+    {
+        public override AccountStateType AccountState => AccountStateType.User;
```
Admin would be good, yes. As long as regular users can't use this command anything can do.

pacampbell commented on Arrowgene.Ddon.GameServer/Handler/InstanceGetEnemySetListHandler.cs:
```diff
@@ -16,8 +16,11 @@ public class InstanceGetEnemySetListHandler : StructurePacketHandler<GameClient,
     {
         private static readonly ServerLogger Logger = LogProvider.Logger<ServerLogger>(typeof(InstanceGetEnemySetListHandler));
 
+        private DdonGameServer _Server;
```
Done.

pacampbell commented on Arrowgene.Ddon.GameServer/Handler/SeasonDungeonGetInfoHandler.cs:
```diff
@@ -0,0 +1,37 @@
+
+            var dungeonInfo = Server.EpitaphRoadManager.GetDungeonInfo(request.DungeonId);
+
+            result.DungeonInfo.Unk0 = 40;
```
Comment for Unk0 or the handler in general?

alborrajo replied:
Just the value and where it comes from. The handler itself is a small part of a way bigger machinery so unless we plant on deeply documenting it all I don't think comments for the handler would be very useful

pacampbell commented on Arrowgene.Ddon.Server/Network/RequestPacketQueueHandler.cs:
I'm going to leave this for now but we can try to do this idea in a separate PR?

alborrajo replied:
Fine by me, no hurries

pacampbell committed:
code review feedback

- Added missing object instantiations to packet consturctors.
- Renamed packets which were not named correctly.
- Added some TODO comments
